first draft sop (#32)
aerogear/mobile-developer-console-operator: SOP/SOP-mdc.adoc

==== To capture the logs

. Capture a snapshot of the 'Mobile Developer Console Application' Grafana dashboard and track it over time. The metrics can be useful for identifying performance issues over time.
. Capture the Service logs
.. Get the pod names by running `oc get pods`
.. Save the logs by running `oc logs <mdc-podname> > <filename>.log` for any pod
. Capture CR Status by running `oc describe MobileDeveloperConsole`.

==== To scale the pod

. Scale the Mobile Develper Console pod
.. Check the number of DESIRED and CURRENT replicas, and make sure that they're the same:
----
oc get deploymentconfig --selector='app=mdc'
NAME      REVISION   DESIRED   CURRENT   TRIGGERED BY
mdc       3          1         1         config,image(mdc-imagestream:1.1.4),image(mdc-oauth-proxy-imagestream:latest)
----
If they're not, this may indicate a problem with the ability to scale up (capacity, missing images, etc), or the ability to scale down (a node where a pod was scheduled might be in trouble etc.).
.. Scale the replication controller if more pods are desired by running `oc scale deploymentconfig --selector='app=mdc' --replicas=X` where X is the number of desired pods.

== Validate

> 

Reviewer: ```suggestion
=== Install
```

. Switch to the Mobile Developer Console namespace by running `oc project <namespace>`  E.g `oc project mobile-developer-console`.
. Check if the mobile developer console pod to is present by running `oc get pods | grep mdc` or `oc get pods | grep mobile-developer-console`.
+
[source,shell]
----
NAME                                           READY     STATUS    RESTARTS   AGE
mdc-58c8877fd8-g6dvr                           1/1       Running   3          9d
----

> -NOTE: The operator is responsible for managing and creating all objects required to have the Mobile Developer Console Service deployed on the cluster.

Reviewer: ```suggestion
NOTE: The operator is responsible for managing and creating all objects required to have the Mobile Developer Console Service deployed on the cluster.
```

-. Check its logs by running `oc logs <operator-podname>`

> -NOTE: You can save the logs by running `oc logs <operator-podname> > <filename>.log`. The logs may provide you with useful information to lead you to the root cause, and they are also useful for providing to the project maintainers when you create an issue.

Reviewer: ```suggestion
NOTE: You can save the logs by running `oc logs <operator-podname> > <filename>.log`. The logs may provide you with useful information to lead you to the root cause, and they are also useful for providing to the project maintainers when you create an issue.
```

.. Check the number of DESIRED and CURRENT replicas, and make sure that they're the same:
----
oc get deploymentconfig --selector='app=mdc'
NAME      REVISION   DESIRED   CURRENT   TRIGGERED BY
mdc       3          1         1         config,image(mdc-imagestream:1.1.4),image(mdc-oauth-proxy-imagestream:latest)
----
If they're not, this may indicate a problem with the ability to scale up (capacity, missing images, etc), or the ability to scale down (a node where a pod was scheduled might be in trouble etc.).
.. Scale the replication controller if more pods are desired by running `oc scale deploymentconfig --selector='app=mdc' --replicas=X` where X is the number of desired pods.

== Validate

=== Install

. Switch to the Mobile Developer Console namespace by running `oc project <namespace>`  E.g `oc project mobile-developer-console`.
. Check if the mobile developer console pod to is present by running `oc get pods | grep mdc` or `oc get pods | grep mobile-developer-console`.
+
[source,shell]
----
NAME                                           READY     STATUS    RESTARTS   AGE
mdc-58c8877fd8-g6dvr                           1/1       Running   3          9d
----

> NOTE: The operator is responsible for managing and creating all objects required to have the Mobile Developer Console Service deployed on the cluster.

Reviewer: Hi @laurafitzgerald shows that is missing steps. Is just a pod created ? It has not ouath, services, route and etc ..?
- Checking the CRs
- Check env variables 
- Check OAuth Proxy Container

etc ...

. Check its logs by running `oc logs <operator-podname>`
NOTE: You can save the logs by running `oc logs <operator-podname> > <filename>.log`. The logs may provide you with useful information to lead you to the root cause, and they are also useful for providing to the project maintainers when you create an issue.

=== Optional configurations

> ==== Monitor

Reviewer: ```suggestion
==== Monitor

If the https://github.com/aerogear/mobile-developer-console-operator#monitoring-service-metrics[Monitoring Service (Metrics)] is enabled for the installation, a Grafana Dashboard titled `MDC Application`, and the Prometheus Monitoring instances are created.
```

Reviewer: Just add the link for the readme.

Author: There is no related heading for this present in README for MDC

Reviewer: https://github.com/aerogear/mobile-developer-console-operator#monitoring-service-metrics[Monitoring Service (Metrics)

Author: wierd. It didn't work for me when I tested it earlier. I'll add it in.

ifdef::env-github[]
:status:
:tip-caption: :bulb:
:note-caption: :information_source:

> :important-caption: :heavy_exclamation_mark:

Reviewer: Missing: 
- Reference Articles
- Prometheus Alerts Procedures
- Validate
-- Installation

See here: https://github.com/aerogear/unifiedpush-operator/blob/master/SOP/SOP-push.adoc

Author: My reasoning for not including the reference article's here is because we include them in the SOP for the operator. The first thing we ask the user to do in the mdc SOP is to check the operator SOP so they would see it there. Do we need to duplicate it?

Author: Same comment as above about the Prometheus Alerts Procedures, seems to be very similar to the overview as it contain text which is very similar to text which appears in other sections.

Author: The text in the push SOP for Prometheus procedures is the following 
```

== Prometheus Alerts Procedures

IMPORTANT: Before checking any of following steps see if the operator pod is running successfully. More info: link:./SOP-operator.adoc[UnifiedPushOperatorDown]

TIP: You can save the logs by running `oc logs <podname> > <filename>.log`. The logs can provided to you useful information in order to identify the root cause of the issue as they are required information to be send to its maintainers in the case it need to be checked by them.
```
To me this doesn't sound like procedure but some general information that we could include elsewhere.

Reviewer: The SOP's need to be done thinking in who has no idea over the project and all. So, it is important IMHO.

:caution-caption: :fire:
:warning-caption: :warning:
:table-caption!:
endif::[]

:toc:
:toc-placement!:

= Mobile Developer Console - Standard Operating Procedures

:toc:
toc::[]

== Overview

The following guide outlines the steps required to manage and solve issues in the https://github.com/aerogear/mobile-developer-console[Mobile Developer Console] which is managed, installed and configured via the https://github.com/aerogear/mobile-developer-console-operator[Mobile Developer Console Operator].

== Success Indicators

All alerts should appear as green in the Prometheus Alert Monitoring.

> IMPORTANT: Before checking any of following steps see if the operator pod is running successfully. More info: https://github.com/aerogear/mobile-developer-console/blob/0.2.0/SOP/SOP-operator.adoc[MobileDeveloperConsoleOperatorDown]

Reviewer 2: This link is broken because it's pointing to the wrong repo.

```suggestion
IMPORTANT: Before checking any of following steps see if the operator pod is running successfully. More info: link:./SOP-operator.adoc[MobileDeveloperConsoleOperatorDown]
```

== Critical 

=== MobileDeveloperConsoleDown or MobileDeveloperConsoleContainerDown

. Switch to the Mobile Developer Console namespace by running `oc project <namespace>`. E.g `oc project mobile-developer-console`.
. Follow the <<Check Operator>> steps.

. Verify the MobileDeveloperConsole is deployed in the same namespace as the operator.
.. Run `oc get MobileDeveloperConsole`
+

> Expected results:

Reviewer: Missing steps for `MobileDeveloperConsolePodCPUHigh `  and `MobileDeveloperConsolePodMemoryHigh `
as to how to scale the pods.

Author: Yes, will add those.

+
.. Run `oc logs <service-podname> -c mdc-oauth-proxy`. E.g `oc logs example-mdc-1-8m5ql -c mdc-oauth-proxy`
+

Logs should include the following:
+
----
2019/08/01 09:39:04 oauthproxy.go:201: mapping path "/" => upstream "http://localhost:4000/"
2019/08/01 09:39:04 oauthproxy.go:228: OAuthProxy configured for  Client ID: mobile-developer-console
2019/08/01 09:39:04 oauthproxy.go:238: Cookie settings: name:_oauth_proxy secure(https):true httponly:false expiry:168h0m0s domain:<default> refresh:disabled
2019/08/01 09:39:04 http.go:56: HTTP: listening on 0.0.0.0:4180
----
+

.. If alternative logs are found in the above step then save the logs by running `oc logs <service-podname> -c mdc-oauth-proxy > <filename>.log`

== Check Operator

. Switch to the Mobile Developer Console namespace by running `oc project <namespace>`  E.g `oc project mobile-developer-console`.
. Check if the operator pod to is present by running `oc get pods | grep operator`.
+
[source,shell]
----
$ oc get pods | grep operator
NAME                                           READY     STATUS    RESTARTS   AGE
mdc-operator-58c8877fd8-g6dvr                  1/1       Running   3          9d
----

+

> NOTE: The operator is responsible for managing and creating all objects required to have the Mobile Developer Console Service and deployed on the cluster.

Reviewer 2: ```suggestion
NOTE: The operator is responsible for managing and creating all objects required to have the Mobile Developer Console Service deployed on the cluster.
```

=== MobileDeveloperConsoleDown or MobileDeveloperConsoleContainerDown

. Switch to the Mobile Developer Console namespace by running `oc project <namespace>`. E.g `oc project mobile-developer-console`.

> . Follow the <<Validate>> steps.

Reviewer: 👍
